Std/Data/List/Basic.lean

```diff
@@ -521,6 +533,29 @@ Split a list at every occurrence of a separator element. The separators are not
 -/
 @[inline] def splitOn [BEq α] (a : α) (as : List α) : List (List α) := as.splitOnP (· == a)
 
+/--
+Split a list at every occurrence of a separator sublist. The separators are not in the result.
+```
+[1, 1, 2, 3, 2, 4, 4].splitOnSublist [2, 3] = [[1, 1], [2, 4, 4]]
+```
+-/
+def splitOnSublist [BEq α] (l sep : List α) : List (List α) := go l sep sep [] where
+  /-- Auxiliary for `splitOnSublist`. The first explicit argument is a list we want to split, the
+  second a separator sublist (`sep`), the third a part of `sep` that we'll use to split the first
+  argument, and the last an accumulated list of elements of the first argument (`acc`). -/
+  go : List α → List α → List α → List α → List (List α)
+  | a::as, sep@(b::bs), [], acc =>
+    if a == b then
+      (acc.reverse.rdropSublist sep)::go as sep bs [a]
+    else
+      (acc.reverse.rdropSublist sep)::go as sep sep [a]
+  | a::as, sep@(_::_), c::cs, acc =>
+    if a == c then
+      go as sep cs (a::acc)
+    else
+      go as sep sep (a::acc)
+  | l, _, _, acc => [acc.reverse++l]
```

Comment:
This implementation is also too complex to be a good "naive algorithm" for specification purposes and not efficient enough to be the "optimized version" of the algorithm. Since this is already intended as specification for `String.splitOn`, my recommendation is to lean into the "naive algorithm" side of things for now and add an "optimized version" later (possibly using KMP):
```lean
def splitOnList [BEq α] (l sep : List α) : List (List α) :=
  if _h : sep.isEmpty then
    [l]
  else
    let rec go (l acc : List α) : List (List α) :=
      if h' : l.isPrefixOf sep then
        have : length l - sep.length < length l := sorry
        acc.reverse :: go (l.drop sep.length) []
      else
        match l with
        | a :: l => go l (a :: acc)
        | [] => h'.elim (by simp [List.isPrefixOf])
    go l []
termination_by go => l.length
```
This is still a bit too imperative for direct usage; I think you actually want to use the lemmas:
```lean
theorem splitOnList_of_infix [DecidableEq α] (hsep : sep ≠ []) (h : sep <:+: l) :
    ∃ s t, s ++ sep ++ t = l ∧ splitOnList l sep = s :: splitOnList t sep := sorry

theorem splitOnList_of_not_infix [DecidableEq α] (hsep : sep ≠ []) (h : ¬sep <:+: l) :
    splitOnList l sep = [l] := sorry
```

Reply:
Actually this might be a better / more useful factoring:
```lean
/-- Returns `(l₁, l₂)` for the first split `l = l₁ ++ l₂` such that `P l₂` returns true. -/
def splitOnceP (P : List α → Bool) (l : List α) : Option (List α × List α) :=
  if P l then
    some ([], l)
  else
    match l with
    | [] => none
    | a :: l => (splitOnceP P l).map fun (l, r) => (a :: l, r)

def splitOnList [BEq α] (l sep : List α) : List (List α) :=
  if _h : sep.isEmpty then
    [l]
  else
    match e : splitOnceP sep.isPrefixOf l with
    | none => [l]
    | some (l₁, l₂) =>
      have : length l₂ - sep.length < length l := sorry
      splitOnList (l₂.drop sep.length) sep
termination_by _ => l.length
```

Reply:
I'll use this implementation.

Reply:
```diff
def splitOnList [BEq α] (l sep : List α) : List (List α) :=
  if _h : sep.isEmpty then
    [l]
  else
    match e : splitOnceP sep.isPrefixOf l with
    | none => [l]
    | some (l₁, l₂) =>
      have : length l₂ - sep.length < length l := sorry
-     splitOnList (l₂.drop sep.length) sep
+     l₁::splitOnList (l₂.drop sep.length) sep
termination_by _ => l.length
```

I fixed the definition as above.